File: mizan-go/webhooks.go
package mizan

import (
	"context"
	"crypto/subtle"
	"encoding/json"
	"errors"
	"fmt"
	"io"
	"math/big"
	"net/http"
	"strings"
)

const (
	// HeaderOutboxID is stable across retries and is the receiver deduplication key.
	HeaderOutboxID = "X-Mizan-Outbox-Id"
	// HeaderAckSequence acknowledges durable application of exactly one ledger sequence.
	HeaderAckSequence = "X-Mizan-Ack-Sequence"
	// defaultWebhookBodyLimit bounds memory use when callers do not configure a limit.
	defaultWebhookBodyLimit = 1 << 20
)

// LedgerEntryType identifies why an immutable ledger entry was committed.
type LedgerEntryType string

const (
	LedgerSubscriptionActivated             LedgerEntryType = "subscription_activated"
	LedgerSubscriptionChangeScheduled       LedgerEntryType = "subscription_change_scheduled"
	LedgerSubscriptionCancellationScheduled LedgerEntryType = "subscription_cancellation_scheduled"
	LedgerSubscriptionCancelled             LedgerEntryType = "subscription_cancelled"
	LedgerSubscriptionCancellationRevoked   LedgerEntryType = "subscription_cancellation_revoked"
	LedgerRenewalFailed                     LedgerEntryType = "renewal_failed"
	LedgerSubscriptionRenewed               LedgerEntryType = "subscription_renewed"
	LedgerAzeerUnitsToppedUp                LedgerEntryType = "azeer_units_topped_up"
	LedgerProviderBalanceToppedUp           LedgerEntryType = "provider_balance_topped_up"
	LedgerProviderBalanceRefunded           LedgerEntryType = "provider_balance_refunded"
	LedgerFeatureBudgetUpdated              LedgerEntryType = "feature_budget_updated"
	LedgerBillingAccountDepleted            LedgerEntryType = "billing_account_depleted"
	LedgerFeaturePausedBudget               LedgerEntryType = "feature_paused_budget"
	LedgerFeaturePausedManual               LedgerEntryType = "feature_paused_manual"
	LedgerFeatureResumedManual              LedgerEntryType = "feature_resumed_manual"
	LedgerUsageRejected                     LedgerEntryType = "usage_rejected"
	LedgerUsageConsumed                     LedgerEntryType = "usage_consumed"
	LedgerIncludedUnitsExpired              LedgerEntryType = "included_units_expired"
	LedgerPurchasedUnitsExpired             LedgerEntryType = "purchased_units_expired"
	LedgerPromotionalUnitsExpired           LedgerEntryType = "promotional_units_expired"
	LedgerPromotionalUnitsGranted           LedgerEntryType = "promotional_units_granted"
	LedgerMonthlyIncludedUnitsGranted       LedgerEntryType = "monthly_included_units_granted"
	LedgerOutboxRetryRequested              LedgerEntryType = "outbox_retry_requested"
)

// NotificationType identifies a non-ledger operational notification.
type NotificationType string

const (
	NotificationBudgetWarning        NotificationType = "budget_warning"
	NotificationBudgetBreached       NotificationType = "budget_breached"
	NotificationBudgetPaused         NotificationType = "budget_paused"
	NotificationFeaturePausedManual  NotificationType = "feature_paused_manual"
	NotificationFeatureResumedManual NotificationType = "feature_resumed_manual"
)

// LedgerEntryWebhook is immutable financial history with versioned effective-time metadata.
type LedgerEntryWebhook struct {
	ID             string          `json:"id"`
	EntryType      LedgerEntryType `json:"entry_type"`
	SourceCommand  string          `json:"source_command"`
	SourceEventID  string          `json:"source_event_id"`
	SubscriptionID *string         `json:"subscription_id,omitempty"`
	FeatureCode    *FeatureCode    `json:"feature_code,omitempty"`
	EffectiveAt    string          `json:"effective_at"`
	CatalogVersion string          `json:"catalog_version"`
	PolicyVersion  string          `json:"policy_version"`
	Metadata       json.RawMessage `json:"metadata"`
}

// LedgerPostingWebhook is one exact debit or credit. Postings balance per rail and unit.
type LedgerPostingWebhook struct {
	Rail        string          `json:"rail"`
	AccountCode string          `json:"account_code"`
	Amount      ExactAmount     `json:"amount"`
	Unit        string          `json:"unit"`
	LotID       *string         `json:"lot_id,omitempty"`
	Metadata    json.RawMessage `json:"metadata,omitempty"`
}

// LedgerWebhook is delivered strictly by BusinessSequence and at least once.
type LedgerWebhook struct {
	EventID          string                 `json:"event_id"`
	BusinessID       string                 `json:"business_id"`
	BusinessSequence int64                  `json:"business_sequence"`
	Entry            LedgerEntryWebhook     `json:"entry"`
	Postings         []LedgerPostingWebhook `json:"postings"`
}

// NotificationWebhook is an at-least-once budget or feature-state notification.
type NotificationWebhook struct {
	Type        NotificationType `json:"type"`
	BusinessID  string           `json:"business_id"`
	FeatureCode FeatureCode      `json:"feature_code"`
	Period      string           `json:"period,omitempty"`
	Projected   ExactAmount      `json:"projected,omitempty"`
	Limit       ExactAmount      `json:"limit,omitempty"`
}

// WebhookContext contains transport identity. OutboxID is stable across retries
// and is the primary key applications should persist for duplicate suppression.
type WebhookContext struct {
	// OutboxID is stable across retries and should be persisted before returning success.
	OutboxID string
	// Headers excludes Authorization so application callbacks cannot leak the bearer token.
	Headers http.Header
	// RawBody is a defensive copy of the exact received JSON bytes.
	RawBody []byte
}

// WebhookHandler receives validated events. Returning nil means the application
// durably accepted the event. Returning an error causes a non-2xx response so
// Mizan retries it; ledger acknowledgement is never emitted on an error.
type WebhookHandler interface {
	HandleLedger(context.Context, LedgerWebhook, WebhookContext) error
	HandleNotification(context.Context, NotificationWebhook, WebhookContext) error
}

// WebhookHandlerFuncs adapts two functions to WebhookHandler.
type WebhookHandlerFuncs struct {
	Ledger       func(context.Context, LedgerWebhook, WebhookContext) error
	Notification func(context.Context, NotificationWebhook, WebhookContext) error
}

// HandleLedger dispatches to Ledger and returns an error when it is not configured.
func (h WebhookHandlerFuncs) HandleLedger(ctx context.Context, event LedgerWebhook, delivery WebhookContext) error {
	if h.Ledger == nil {
		return errors.New("mizan: ledger webhook handler is not configured")
	}
	return h.Ledger(ctx, event, delivery)
}

// HandleNotification dispatches to Notification and returns an error when it is not configured.
func (h WebhookHandlerFuncs) HandleNotification(ctx context.Context, event NotificationWebhook, delivery WebhookContext) error {
	if h.Notification == nil {
		return errors.New("mizan: notification webhook handler is not configured")
	}
	return h.Notification(ctx, event, delivery)
}

// WebhookResponse is framework-neutral. WriteHTTP can be used with net/http;
// the Fiber adapter maps the same response without changing protocol behavior.
type WebhookResponse struct {
	StatusCode int
	Headers    http.Header
	Body       []byte
}

// WriteHTTP writes the framework-neutral status, headers, and body to net/http.
func (r WebhookResponse) WriteHTTP(w http.ResponseWriter) {
	for name, values := range r.Headers {
		for _, value := range values {
			w.Header().Add(name, value)
		}
	}
	w.WriteHeader(r.StatusCode)
	_, _ = w.Write(r.Body)
}

// WebhookReceiver validates and dispatches both Mizan webhook streams. Set
// BearerToken when the delivery endpoint is configured with bearer auth.
type WebhookReceiver struct {
	Handler      WebhookHandler
	BearerToken  string
	MaxBodyBytes int
}

// ServeHTTP receives both webhook streams as a standard net/http handler.
func (r WebhookReceiver) ServeHTTP(w http.ResponseWriter, request *http.Request) {
	limit := r.bodyLimit()
	// Read one bounded body before delegating to the framework-neutral receiver.
	body, err := readBoundedBody(request, limit)
	if err != nil {
		r.errorResponse(http.StatusRequestEntityTooLarge, "request body exceeds the configured limit").WriteHTTP(w)
		return
	}
	r.Receive(request.Context(), request.Header, body).WriteHTTP(w)
}

// Receive lets any framework pass request headers and raw JSON to the SDK.
func (r WebhookReceiver) Receive(ctx context.Context, headers http.Header, body []byte) WebhookResponse {
	if len(body) == 0 {
		return r.errorResponse(http.StatusBadRequest, "request body is required")
	}
	if len(body) > r.bodyLimit() {
		return r.errorResponse(http.StatusRequestEntityTooLarge, "request body exceeds the configured limit")
	}
	if !r.authorized(headers.Get("Authorization")) {
		return r.errorResponse(http.StatusUnauthorized, "webhook authorization failed")
	}
	outboxID := strings.TrimSpace(headers.Get(HeaderOutboxID))
	if outboxID == "" {
		return r.errorResponse(http.StatusBadRequest, HeaderOutboxID+" is required")
	}
	if r.Handler == nil {
		return r.errorResponse(http.StatusServiceUnavailable, "webhook handler is not configured")
	}

	var envelope map[string]json.RawMessage
	// Decode only the top-level keys first so stream type can be selected safely.
	if err := json.Unmarshal(body, &envelope); err != nil || envelope == nil {
		return r.errorResponse(http.StatusBadRequest, "request body must be a JSON object")
	}
	callbackHeaders := headers.Clone()
	// Never expose the bearer credential to application handlers or their logs.
	callbackHeaders.Del("Authorization")
	delivery := WebhookContext{OutboxID: outboxID, Headers: callbackHeaders, RawBody: append([]byte(nil), body...)}
	if _, ledger := envelope["business_sequence"]; ledger {
		// Ledger shape takes precedence and is acknowledged only after handler success.
		var event LedgerWebhook
		if err := json.Unmarshal(body, &event); err != nil {
			return r.errorResponse(http.StatusUnprocessableEntity, "ledger webhook does not match the contract")
		}
		if err := validateLedgerWebhook(event); err != nil {
			return r.errorResponse(http.StatusUnprocessableEntity, err.Error())
		}
		if err := r.Handler.HandleLedger(ctx, event, delivery); err != nil {
			return r.errorResponse(http.StatusInternalServerError, "ledger webhook processing failed")
		}
		return WebhookResponse{StatusCode: http.StatusNoContent, Headers: http.Header{HeaderAckSequence: []string{fmt.Sprint(event.BusinessSequence)}}}
	}
	if _, notification := envelope["type"]; notification {
		// Notifications use outbox deduplication but have no sequence acknowledgement.
		var event NotificationWebhook
		if err := json.Unmarshal(body, &event); err != nil {
			return r.errorResponse(http.StatusUnprocessableEntity, "notification webhook does not match the contract")
		}
		if err := validateNotificationWebhook(event); err != nil {
			return r.errorResponse(http.StatusUnprocessableEntity, err.Error())
		}
		if err := r.Handler.HandleNotification(ctx, event, delivery); err != nil {
			return r.errorResponse(http.StatusInternalServerError, "notification webhook processing failed")
		}
		return WebhookResponse{StatusCode: http.StatusNoContent, Headers: make(http.Header)}
	}
	return r.errorResponse(http.StatusUnprocessableEntity, "unknown Mizan webhook payload")
}

func (r WebhookReceiver) bodyLimit() int {
	if r.MaxBodyBytes > 0 {
		return r.MaxBodyBytes
	}
	return defaultWebhookBodyLimit
}

func (r WebhookReceiver) authorized(value string) bool {
	if r.BearerToken == "" {
		return true
	}
	expected := "Bearer " + r.BearerToken
	// Equal-length constant-time comparison avoids token-dependent timing differences.
	return len(value) == len(expected) && subtle.ConstantTimeCompare([]byte(value), []byte(expected)) == 1
}

func (r WebhookReceiver) errorResponse(status int, message string) WebhookResponse {
	body, _ := json.Marshal(map[string]any{"error": map[string]string{"message": message}})
	return WebhookResponse{StatusCode: status, Headers: http.Header{"Content-Type": []string{"application/json"}}, Body: body}
}

func validateLedgerWebhook(event LedgerWebhook) error {
	if event.EventID == "" || event.Entry.ID == "" || event.EventID != event.Entry.ID {
		return errors.New("event_id must be present and equal entry.id")
	}
	if event.BusinessID == "" || event.BusinessSequence < 1 {
		return errors.New("business_id and a positive business_sequence are required")
	}
	if !validLedgerEntryType(event.Entry.EntryType) || event.Entry.SourceCommand == "" || event.Entry.SourceEventID == "" || event.Entry.EffectiveAt == "" || event.Entry.CatalogVersion == "" || event.Entry.PolicyVersion == "" || len(event.Entry.Metadata) == 0 {
		return errors.New("ledger entry fields do not match the contract")
	}
	balances := map[string]*big.Int{}
	for _, posting := range event.Postings {
		if posting.AccountCode == "" || (posting.Rail != "azeer_units" && posting.Rail != "provider_balance" && posting.Rail != "invoice") || (posting.Unit != "milliunit" && posting.Unit != "halala") {
			return errors.New("ledger posting fields do not match the contract")
		}
		amount, ok := new(big.Int).SetString(string(posting.Amount), 10)
		if !ok || !amount.IsInt64() || !isCanonicalInteger(string(posting.Amount)) {
			return errors.New("ledger posting amounts must be exact integer strings")
		}
		key := posting.Rail + "\x00" + posting.Unit
		// Rail and unit form separate balancing domains; money cannot offset milliunits.
		if balances[key] == nil {
			balances[key] = new(big.Int)
		}
		balances[key].Add(balances[key], amount)
	}
	for _, balance := range balances {
		if balance.Sign() != 0 {
			return errors.New("ledger postings must balance to zero per rail and unit")
		}
	}
	return nil
}

func validateNotificationWebhook(event NotificationWebhook) error {
	if event.BusinessID == "" || event.FeatureCode == "" {
		return errors.New("notification business_id and feature_code are required")
	}
	switch event.Type {
	case NotificationBudgetWarning, NotificationBudgetBreached:
		if event.Period == "" || !isUnsignedExactAmount(string(event.Projected)) || !isUnsignedExactAmount(string(event.Limit)) {
			return errors.New("budget threshold notifications require period, projected, and limit")
		}
	case NotificationBudgetPaused:
		if event.Period == "" {
			return errors.New("budget_paused notifications require period")
		}
	case NotificationFeaturePausedManual, NotificationFeatureResumedManual:
	default:
		return errors.New("notification type is not supported")
	}
	return nil
}

func validLedgerEntryType(value LedgerEntryType) bool {
	switch value {
	case LedgerSubscriptionActivated, LedgerSubscriptionChangeScheduled, LedgerSubscriptionCancellationScheduled,
		LedgerSubscriptionCancelled, LedgerSubscriptionCancellationRevoked, LedgerRenewalFailed,
		LedgerSubscriptionRenewed, LedgerAzeerUnitsToppedUp, LedgerProviderBalanceToppedUp,
		LedgerProviderBalanceRefunded, LedgerFeatureBudgetUpdated, LedgerBillingAccountDepleted,
		LedgerFeaturePausedBudget, LedgerFeaturePausedManual, LedgerFeatureResumedManual,
		LedgerUsageRejected, LedgerUsageConsumed, LedgerIncludedUnitsExpired, LedgerPurchasedUnitsExpired,
		LedgerPromotionalUnitsExpired, LedgerPromotionalUnitsGranted, LedgerMonthlyIncludedUnitsGranted,
		LedgerOutboxRetryRequested:
		return true
	default:
		return false
	}
}

func isCanonicalInteger(value string) bool {
	if strings.HasPrefix(value, "-") {
		return len(value) > 1 && isUnsignedCanonicalInteger(value[1:])
	}
	return isUnsignedCanonicalInteger(value)
}

func isUnsignedCanonicalInteger(value string) bool {
	if value == "0" {
		return true
	}
	if value == "" || value[0] < '1' || value[0] > '9' {
		return false
	}
	for i := 1; i < len(value); i++ {
		if value[i] < '0' || value[i] > '9' {
			return false
		}
	}
	return true
}

func isUnsignedExactAmount(value string) bool {
	amount, ok := new(big.Int).SetString(value, 10)
	return ok && amount.IsInt64() && amount.Sign() >= 0 && isUnsignedCanonicalInteger(value)
}

func readBoundedBody(request *http.Request, limit int) ([]byte, error) {
	defer request.Body.Close()
	// Read one extra byte so the exact limit and an oversized body are distinguishable.
	body, err := io.ReadAll(io.LimitReader(request.Body, int64(limit)+1))
	if err != nil {
		return nil, err
	}
	if len(body) > limit {
		return nil, errors.New("body too large")
	}
	return body, nil
}

File: mizan-python/src/mizan/webhooks.py
"""Framework-neutral receivers for Mizan ledger and notification webhooks."""

from __future__ import annotations

import hmac
import inspect
import json
from dataclasses import dataclass
from typing import Any, Awaitable, Callable, Literal, Mapping, TypeAlias, TypedDict, cast

from .models import ExactAmount

# Stable delivery identity; persist this value to suppress at-least-once duplicates.
OUTBOX_ID_HEADER = "x-mizan-outbox-id"
# Ledger receivers emit this header only after durable application success.
ACK_SEQUENCE_HEADER = "x-mizan-ack-sequence"
# Default protects framework adapters from unbounded request bodies.
DEFAULT_MAX_BODY_BYTES = 1 << 20

LedgerEntryType: TypeAlias = Literal[
    "subscription_activated",
    "subscription_change_scheduled",
    "subscription_cancellation_scheduled",
    "subscription_cancelled",
    "subscription_cancellation_revoked",
    "renewal_failed",
    "subscription_renewed",
    "azeer_units_topped_up",
    "provider_balance_topped_up",
    "provider_balance_refunded",
    "feature_budget_updated",
    "billing_account_depleted",
    "feature_paused_budget",
    "feature_paused_manual",
    "feature_resumed_manual",
    "usage_rejected",
    "usage_consumed",
    "included_units_expired",
    "purchased_units_expired",
    "promotional_units_expired",
    "promotional_units_granted",
    "monthly_included_units_granted",
    "outbox_retry_requested",
]

NotificationType: TypeAlias = Literal[
    "budget_warning",
    "budget_breached",
    "budget_paused",
    "feature_paused_manual",
    "feature_resumed_manual",
]


class LedgerEntryWebhook(TypedDict):
    """Immutable business-ledger entry with versioned effective-time metadata."""
    id: str
    entry_type: LedgerEntryType
    source_command: str
    source_event_id: str
    subscription_id: str | None
    feature_code: str | None
    effective_at: str
    catalog_version: str
    policy_version: str
    metadata: dict[str, Any]


class LedgerPostingWebhook(TypedDict, total=False):
    """One exact debit or credit; postings balance independently per rail and unit."""
    rail: Literal["azeer_units", "provider_balance", "invoice"]
    account_code: str
    amount: ExactAmount
    unit: Literal["milliunit", "halala"]
    lot_id: str | None
    metadata: dict[str, Any]


class LedgerWebhook(TypedDict):
    """Strictly ordered ledger delivery; deduplicate by outbox ID and event ID."""
    event_id: str
    business_id: str
    business_sequence: int
    entry: LedgerEntryWebhook
    postings: list[LedgerPostingWebhook]


class NotificationWebhook(TypedDict, total=False):
    """At-least-once operational notification without ledger ordering semantics."""
    type: NotificationType
    business_id: str
    feature_code: str
    period: str
    projected: ExactAmount
    limit: ExactAmount


@dataclass(frozen=True)
class WebhookContext:
    """Transport identity passed to application callbacks.

    ``outbox_id`` is stable across retries and should be the application's
    durable duplicate-suppression key. ``raw_body`` is retained for applications
    that archive the exact received contract.
    """

    outbox_id: str
    headers: Mapping[str, str]
    raw_body: bytes


@dataclass(frozen=True)
class WebhookResponse:
    """Framework-neutral HTTP response returned by :class:`WebhookReceiver`."""
    status_code: int
    headers: Mapping[str, str]
    body: bytes = b""


LedgerCallback: TypeAlias = Callable[[LedgerWebhook, WebhookContext], None | Awaitable[None]]
NotificationCallback: TypeAlias = Callable[[NotificationWebhook, WebhookContext], None | Awaitable[None]]
# Framework-neutral input supports raw bytes/strings and already-decoded test integrations.
WebhookPayload: TypeAlias = bytes | bytearray | str | Mapping[str, Any]


class WebhookReceiver:
    """Authenticate, validate, dispatch, and acknowledge both webhook streams.

    A callback must return only after its application effect has been durably
    committed. Raising an exception produces HTTP 500 without a ledger sequence
    acknowledgement, causing Mizan to retry the same outbox item.
    """

    def __init__(
        self,
        *,
        on_ledger: LedgerCallback,
        on_notification: NotificationCallback,
        bearer_token: str | None = None,
        max_body_bytes: int = DEFAULT_MAX_BODY_BYTES,
    ) -> None:
        if not callable(on_ledger) or not callable(on_notification):
            raise TypeError("on_ledger and on_notification must be callable")
        if bearer_token == "":
            raise ValueError("bearer_token must not be empty")
        if max_body_bytes < 1:
            raise ValueError("max_body_bytes must be positive")
        self._on_ledger = on_ledger
        self._on_notification = on_notification
        self._bearer_token = bearer_token
        self._max_body_bytes = max_body_bytes

    async def receive(self, headers: Mapping[str, str], payload: WebhookPayload) -> WebhookResponse:
        """Receive headers and raw JSON (or an already decoded JSON object).

        This is the integration point for any framework. Fiber and FastAPI
        adapters call the same method and therefore share identical behavior.
        """

        normalized_headers = {str(name).lower(): str(value) for name, value in headers.items()}
        # Authenticate before parsing so unauthorized callers cannot exercise the JSON contract.
        if not self._authorized(normalized_headers.get("authorization")):
            return _error_response(401, "webhook authorization failed")
        outbox_id = normalized_headers.get(OUTBOX_ID_HEADER, "").strip()
        if not outbox_id:
            return _error_response(400, "X-Mizan-Outbox-Id is required")

        try:
            raw_body, decoded = _decode_payload(payload, self._max_body_bytes)
        except _PayloadError as error:
            return _error_response(error.status_code, str(error))
        # Never forward the bearer credential to application callbacks or logging code.
        callback_headers = {name: value for name, value in normalized_headers.items() if name != "authorization"}
        context = WebhookContext(outbox_id=outbox_id, headers=callback_headers, raw_body=raw_body)

        if "business_sequence" in decoded:
            # Ledger shape takes precedence and requires ordered acknowledgement after callback success.
            try:
                event = _validate_ledger(decoded)
            except ValueError as error:
                return _error_response(422, str(error))
            try:
                result = self._on_ledger(event, context)
                if inspect.isawaitable(result):
                    await result
            except Exception:
                return _error_response(500, "ledger webhook processing failed")
            return WebhookResponse(204, {ACK_SEQUENCE_HEADER: str(event["business_sequence"])})

        if "type" in decoded:
            # Notifications are deduplicated by outbox ID but do not acknowledge a ledger sequence.
            try:
                notification = _validate_notification(decoded)
            except ValueError as error:
                return _error_response(422, str(error))
            try:
                result = self._on_notification(notification, context)
                if inspect.isawaitable(result):
                    await result
            except Exception:
                return _error_response(500, "notification webhook processing failed")
            return WebhookResponse(204, {})

        return _error_response(422, "unknown Mizan webhook payload")

    def _authorized(self, authorization: str | None) -> bool:
        if self._bearer_token is None:
            return True
        # Constant-time comparison avoids leaking a configured bearer token by timing.
        return authorization is not None and hmac.compare_digest(
            authorization.encode("utf-8"), f"Bearer {self._bearer_token}".encode("utf-8")
        )


class _PayloadError(ValueError):
    def __init__(self, status_code: int, message: str) -> None:
        super().__init__(message)
        self.status_code = status_code


def _decode_payload(payload: WebhookPayload, limit: int) -> tuple[bytes, dict[str, Any]]:
    """Normalize supported payload forms while preserving exact raw bytes for callbacks."""
    if isinstance(payload, Mapping):
        raw = json.dumps(payload, separators=(",", ":"), ensure_ascii=False).encode("utf-8")
        decoded: Any = dict(payload)
    else:
        raw = payload.encode("utf-8") if isinstance(payload, str) else bytes(payload)
        if not raw:
            raise _PayloadError(400, "request body is required")
        try:
            decoded = json.loads(raw)
        except (UnicodeDecodeError, json.JSONDecodeError) as error:
            raise _PayloadError(400, "request body must be a JSON object") from error
    if len(raw) > limit:
        raise _PayloadError(413, "request body exceeds the configured limit")
    if not isinstance(decoded, dict):
        raise _PayloadError(400, "request body must be a JSON object")
    return raw, cast(dict[str, Any], decoded)


# Closed entry vocabulary prevents silently accepting a ledger reason the SDK cannot interpret.
_LEDGER_TYPES = {
    "subscription_activated", "subscription_change_scheduled", "subscription_cancellation_scheduled",
    "subscription_cancelled", "subscription_cancellation_revoked", "renewal_failed",
    "subscription_renewed", "azeer_units_topped_up", "provider_balance_topped_up",
    "provider_balance_refunded", "feature_budget_updated", "billing_account_depleted",
    "feature_paused_budget", "feature_paused_manual", "feature_resumed_manual", "usage_rejected",
    "usage_consumed", "included_units_expired", "purchased_units_expired",
    "promotional_units_expired", "promotional_units_granted", "monthly_included_units_granted",
    "outbox_retry_requested",
}


def _validate_ledger(payload: dict[str, Any]) -> LedgerWebhook:
    """Validate ledger identity, exact postings, and independent rail/unit balance."""
    event_id = payload.get("event_id")
    business_id = payload.get("business_id")
    sequence = payload.get("business_sequence")
    entry = payload.get("entry")
    postings = payload.get("postings")
    if not isinstance(event_id, str) or not event_id or not isinstance(business_id, str) or not business_id:
        raise ValueError("ledger event_id and business_id are required")
    if isinstance(sequence, bool) or not isinstance(sequence, int) or sequence < 1:
        raise ValueError("business_sequence must be a positive integer")
    if not isinstance(entry, dict) or entry.get("id") != event_id:
        raise ValueError("event_id must be equal to entry.id")
    required_entry_strings = (
        "id", "entry_type", "source_command", "source_event_id", "effective_at", "catalog_version", "policy_version"
    )
    if any(not isinstance(entry.get(field), str) or not entry[field] for field in required_entry_strings):
        raise ValueError("ledger entry fields do not match the contract")
    if entry["entry_type"] not in _LEDGER_TYPES or not isinstance(entry.get("metadata"), dict):
        raise ValueError("ledger entry fields do not match the contract")
    if not isinstance(postings, list):
        raise ValueError("ledger postings must be an array")
    balances: dict[tuple[str, str], int] = {}
    for posting in postings:
        if not isinstance(posting, dict):
            raise ValueError("ledger posting fields do not match the contract")
        rail, unit, amount = posting.get("rail"), posting.get("unit"), posting.get("amount")
        if rail not in {"azeer_units", "provider_balance", "invoice"} or unit not in {"milliunit", "halala"}:
            raise ValueError("ledger posting fields do not match the contract")
        if not isinstance(posting.get("account_code"), str) or not posting["account_code"] or not _canonical_integer(amount):
            raise ValueError("ledger posting amounts must be exact integer strings")
        key = (rail, unit)
        # Money and unit rails must each balance independently to zero.
        balances[key] = balances.get(key, 0) + int(cast(str, amount))
    if any(balance != 0 for balance in balances.values()):
        raise ValueError("ledger postings must balance to zero per rail and unit")
    return cast(LedgerWebhook, payload)


def _validate_notification(payload: dict[str, Any]) -> NotificationWebhook:
    """Validate type-dependent notification fields and exact threshold amounts."""
    event_type = payload.get("type")
    if not isinstance(payload.get("business_id"), str) or not payload["business_id"]:
        raise ValueError("notification business_id and feature_code are required")
    if not isinstance(payload.get("feature_code"), str) or not payload["feature_code"]:
        raise ValueError("notification business_id and feature_code are required")
    if event_type in {"budget_warning", "budget_breached"}:
        if not isinstance(payload.get("period"), str) or not payload["period"]:
            raise ValueError("budget threshold notifications require period, projected, and limit")
        if not _unsigned_canonical_integer(payload.get("projected")) or not _unsigned_canonical_integer(payload.get("limit")):
            raise ValueError("budget threshold notifications require period, projected, and limit")
    elif event_type == "budget_paused":
        if not isinstance(payload.get("period"), str) or not payload["period"]:
            raise ValueError("budget_paused notifications require period")
    elif event_type not in {"feature_paused_manual", "feature_resumed_manual"}:
        raise ValueError("notification type is not supported")
    return cast(NotificationWebhook, payload)


def _canonical_integer(value: Any) -> bool:
    """Accept only canonical signed-int64 strings without leading zeroes."""
    if not isinstance(value, str) or not value:
        return False
    canonical = _unsigned_canonical_digits(value[1:]) if value.startswith("-") else _unsigned_canonical_digits(value)
    return canonical and -(2**63) <= int(value) <= 2**63 - 1


def _unsigned_canonical_integer(value: Any) -> bool:
    """Accept only canonical non-negative int64 strings."""
    return isinstance(value, str) and _unsigned_canonical_digits(value) and int(value) <= 2**63 - 1


def _unsigned_canonical_digits(value: str) -> bool:
    return value == "0" or (value[0:1] in "123456789" and value.isdigit())


def _error_response(status_code: int, message: str) -> WebhookResponse:
    body = json.dumps({"error": {"message": message}}, separators=(",", ":")).encode("utf-8")
    return WebhookResponse(status_code, {"content-type": "application/json"}, body)
